commit bec5df6e4278dec8b9d592b2f93930327a2b8279
Author: q66 <q66@chimera-linux.org>
Date:   Sat Nov 11 02:41:10 2023 +0100

    steamcompmgr: avoid xwayland server mutex deadlock
    
    This uses pthread_exit, which does not ensure C++ destructors
    are called. That means the steamcompmgr mutex may remain locked
    forever, which will prevent gamescope from quitting, as wlserver
    will deadlock on trying to lock it.
    
    Take this strategy: when steamcompmgr_main exits normally, there
    is no need to call pthread_exit as the thread will stop on its
    own via returning from its function; this will correctly release
    all resources.
    
    Upon invocation of error handler however, do call pthread_exit
    and unlock the mutex manually.

--- a/src/steamcompmgr.cpp
+++ b/src/steamcompmgr.cpp
@@ -782,6 +782,7 @@ extern float g_flMaxWindowScale;
 bool			synchronize;
 
 std::mutex g_SteamCompMgrXWaylandServerMutex;
+static std::unique_lock<std::mutex> *g_SteamCompMgrXWaylandServerLock;
 
 gamescope::VBlankTime g_SteamCompMgrVBlankTime = {};
 
@@ -5895,8 +5896,8 @@ error(Display *dpy, XErrorEvent *ev)
 	return 0;
 }
 
-[[noreturn]] static void
-steamcompmgr_exit(void)
+static void
+steamcompmgr_exit(bool exit)
 {
 	g_ImageWaiter.Shutdown();
 
@@ -5937,14 +5938,19 @@ steamcompmgr_exit(void)
     wlserver_shutdown();
     wlserver_unlock(false);
 
-	pthread_exit(NULL);
+	if (exit)
+	{
+		// pthread_exit may not call C++ destructors
+		g_SteamCompMgrXWaylandServerLock->unlock();
+		pthread_exit(NULL);
+	}
 }
 
-static int
+[[noreturn]] static int
 handle_io_error(Display *dpy)
 {
 	xwm_log.errorf("X11 I/O error");
-	steamcompmgr_exit();
+	steamcompmgr_exit(true);
 }
 
 static bool
@@ -7495,6 +7501,7 @@ steamcompmgr_main(int argc, char **argv)
 	init_runtime_info();
 
 	std::unique_lock<std::mutex> xwayland_server_guard(g_SteamCompMgrXWaylandServerMutex);
+	g_SteamCompMgrXWaylandServerLock = &xwayland_server_guard;
 
 	// Initialize any xwayland ctxs we have
 	{
@@ -8037,7 +8044,7 @@ steamcompmgr_main(int argc, char **argv)
 		vblank = false;
 	}
 
-	steamcompmgr_exit();
+	steamcompmgr_exit(false);
 }
 
 void steamcompmgr_send_frame_done_to_focus_window()
